design/0005-circuit-breaking.md

```diff
@@ -0,0 +1,231 @@
+# Title
+
+- Contribution Name: circuit-breaking
+- Implementation Owner: @Ashish-Bansal, Linkerd Team
+- Start Date: 2020-06-14
+- Target Date: N.A.
+- Linkerd Issue:
+  [linkerd/linkerd2#2846](https://github.com/linkerd/linkerd2/issues/2846)
+- Reviewers: @adleong
+
+## Summary
+
+[summary]: #summary
+
+We would like to implement circuit breaking functionality which would allow end
+users to configure mesh such that they can limit the impact of failures from
+faulty service over others and prevent cascading issues. By configuring circuit
+breaking, they will be able to remove failing service instances from load
+balancing and survive through it in a more graceful manner.
+
+## Problem Statement
+
+In a service mesh, services often make calls to other services. In case upstream
+service is unable to respond due to any reason, it's possible that these
+failures can cascade to other services in the cluster.
+
+If we are aware that specific service instance is not responding, we can call it
+unhealthy and it's better if we just stop sending "traffic"[0] to that instance
+rather than getting just another expected failure.
+
+---
+
+It would also be somewhat helpful in case of canary deployments when there's
+some regression and new build is raising failures.
+
+[Example](https://drive.google.com/open?id=1FxKi-egyhHumCvlHvTI4P114z7qOLPeN)
+
+---
+
+Users should also be able to set max-ejection-percentage which provides max
```

Comment:
Hm, I got it from the envoy and it was more of a, ahem ahem, blind copying. 

To think about it again, it means we will keep sending traffic to the upstream even when the circuit breaker should be in open state for those upstreams. So, in a way, it can act as a safe default. For example, let's say max-ejection-percentage is 10%, user can modify all other configuration options related to circuit-breaker and see how those settings affect the cluster in different scenarios. This way they won't end up taking down their whole service due to the same. But yeah, I'm not sure if it's convincing enough to have it for such a usecase -- well it's not for my own personal use case.